Merge duality neutron (CORRECT BRANCH)

jcompagni10 opened the pull request:
Merge duality modules (dex, epochs, gmp, ibcswap & incentives) into neutron.

Commit:
copy over duality files

Commit:
update duality code to connect with neutron

Commit:
downgrade packet-forward-middleware

the version of packet-forward-middleware being used by neutron breaks duality's
swap-and-forward middleware due to a change in how receiever addresses are
handled.

Commit:
add dex, incentives and epochs modules to storeUpgrades keys

Commit:
change dex params to kv store

Commit:
edit and recompile protos to match neutron namespace

Commit:
switch incentives to kvstore params

Commit:
start using "cosmossdk.io/math"

Commit:
FIX: deprecation github.com/cosmos/cosmos-sdk/types/errors => cosmossdk.io/errors

Commit:
FIX: missing import

Commit:
misc linting fixes

Commit:
go fumpt --extra on all duality files

Commit:
General Test cleanup

cleanup KeeperTestHelper and re-use for dex testing

centralize event testing assertions

rename test suite to be more informative

Commit:
set max_gas in init-neutrond.sh

duality code requires max gas to be set initChain panics

Commit:
remove vestigial code

Commit:
rename duality => neutron where applicable in code

Commit:
add dex and incentives updaateParams msgs to proposal whitelist

Commit:
Add note to dex and incentives UpdateParams msg

Commit:
FIX: update incentives cli tests to match new denom format

Commit:
FIX: more canonical naming for GoodTilPurgeHitGasLimit event

Commit:
Audit Fix: Don't use coin.Add in deposit

every sdk.Coin.Add happening in DepositCore results in multiple traversals over all coins
in sharesIssued , and also involves allocation of additional data structures, thus resulting in quadratic
scaling of the DepositCore runtime (because it invokes sdk.Coins Add() for each issued share).

Commit:
Mint shares at the end depositCore; not inside the loop

Much better gas efficiency since we don't need to perform duplicate lookups for
each share issuance

Also add additional check for duplicate deposits. This should exist regardless
but also required for share math to work if we aren't issuing shares within the
deposit loop

Commit:
Audit Fix: check that deposit options array is correct length

Commit:
FIX: remove vestigial tests

Commit:
Rewrite test helpers to default to BIGToken (exponent 6)

In order to simulate more realistic trade volume and avoid inadvertent failures due to ErrInvalidPositionSpread
all of the basic user operations (fundXXXBalance, assertXXXBalance, XXXLimitsSells, etc.) treat TokenA and TokenB
as BIG tokens with an exponent of 6. Ie. fundAliceBalance(10, 10) funds alices account with 10,000,000 small TokenA and TokenB.
For tests requiring more accuracy methods that take Ints (ie. assertXXXAccountBalancesInt, NewWithdrawlInt) are used
and assume that amount are being provided in terms of small tokens.

Example:
s.fundAliceBalances(10, 10)
s.assertAliceBalances(10, 10) ==> True
s.assertAliceBalancesInt(sdkmath.NewInt(10_000_000), sdkmath.NewInt(10_000_000)) ==> true

Commit:
small fixes per PR feedback

-typos,
-remove extra maccperms
-remove unused expected keepers
-incorrect module for burning coins for ibcswap
-adding minting restriction for dex

Commit:
remove outdated READMEs from code ported from osmosis

swelf19 commented on x/dex/types/tx.pb.go:
```diff
@@ -0,0 +1,5005 @@
+	Receiver        string                                   `protobuf:"bytes,2,opt,name=receiver,proto3" json:"receiver,omitempty"`
+	TokenA          string                                   `protobuf:"bytes,3,opt,name=tokenA,proto3" json:"tokenA,omitempty"`
+	TokenB          string                                   `protobuf:"bytes,4,opt,name=tokenB,proto3" json:"tokenB,omitempty"`
+	AmountsA        []github_com_cosmos_cosmos_sdk_types.Int `protobuf:"bytes,5,rep,name=amountsA,proto3,customtype=github.com/cosmos/cosmos-sdk/types.Int" json:"amountA" yaml:"amountsA"`
```
go struct name - Amount**s**A
proto field name - amount**s**A
and json field name - amountA (singular)
Isnt it a good time to make names consistent

jcompagni10 replied:
👍

pr0n00gler commented:
https://github.com/neutron-org/neutron-tests/actions/runs/6734653064

Commit:
typo: fix inconsistent dex/tx.proto naming

Commit:
Remove incentives and epochs module.

The rationale is as follows:
1. No direct need to incentivize the base pools right away. Most of that is done on astroport and integrations
2. The modules were modified from Osmo incentives and the degree of certainty we have here (even after 2 audits) is not the same as the core Dex module.
3. Upside of incentives is low compared to risk of pushing sub-optimal code

Given that I think we should hold off on the deployment of the incentives, keep it back here for a while longer for further testing etc and decide if we really want it in the future.

pr0n00gler committed:
small improvements

pr0n00gler committed:
fix core.go

pr0n00gler committed:
pool validation

NeverHappened committed:
fix: add more multihop validation

pr0n00gler committed:
Merge pull request #353 from neutron-org/remove_incentives_module

Remove incentives and epochs module.

pr0n00gler committed:
Merge branch 'merge_duality_neutron' into fix/small-improvements-duality

pr0n00gler committed:
camelCase -> snake_case

pr0n00gler committed:
fix json and yaml tags

pr0n00gler committed:
fix json tags

Commit:
save

Commit:
Merge branch 'update-sdk47' into merge_duality_neutron

pr0n00gler committed:
fix max_gas limit for tests

pr0n00gler commented:
https://github.com/neutron-org/neutron-tests/actions/runs/6802727604

Commit:
Upgrade x/ibcswap to be compatible with PFM v7.0.2

- Fix issues relating to PFMs new overrideReceiver address
- Small cleanup up OnRecvPacket for greater clarity
- Get rid of NonRefundable swap param and just use presence NeutronRefundAddress
  to determine refund behavior.
- Improve refund behavior to allow for IBC refund after failed forward
- Add additional validation to ensure that only the amount transfered is used
  for swap

Commit:
switch rounding to favor taker

update tests accordingly

Commit:
Ensure deposit, withdraw and place/withdraw limit order correctly validate
inputs to ensure that abs(tick) + fee < MaxTickExp

This not only provides better validation for transactins that will fail at later
price checks, but also ensures that tick + fee combinations that could overflow an
int64 are not allowed into the system.

Commit:
Add extra logic to ensure GetLimitOrderTrancheByKey cannot return nil, true

Commit:
Rename SwapExactAmountIn to be more accurate and add more disclaimers

Commit:
FIX: valid options array is correct length for deposit

Commit:
Add swap logic to abort if taker will be getting an unfair price due to rounding

Commit:
handle swap case where amountIn is 0

Commit:
add tests for swap rouding behavior

pr0n00gler committed:
revert returning an error in Swap()

pr0n00gler committed:
Merge pull request #343 from neutron-org/fix_deposit_validation

Audit Fix: deposit msg validation

pr0n00gler committed:
Merge pull request #356 from neutron-org/fix/incomplete-pool-validation

Fix: incomplete pool validation

pr0n00gler committed:
Merge pull request #355 from neutron-org/fix/small-improvements-duality

Fix: small improvements

pr0n00gler committed:
Merge branch 'merge_duality_neutron' into fix_coin_add

pr0n00gler committed:
Merge pull request #342 from neutron-org/fix_coin_add

Audit Fix: Fix coins.add usage during deposit

pr0n00gler committed:
Merge branch 'merge_duality_neutron' into fix/msg-multihop-validation

pr0n00gler committed:
Merge branch 'merge_duality_neutron' into fix/misc_security_fixes

pr0n00gler committed:
Merge pull request #361 from neutron-org/fix/misc_security_fixes

Fix/misc security fixes

Commit:
Store MaxTrueTakerSpread as a dex param

Commit:
FIX: cleanup tests; make linter happy

pr0n00gler committed:
Merge pull request #357 from neutron-org/fix/msg-multihop-validation

fix: msg multihop validation #ntrn-157

pr0n00gler committed:
Merge pull request #360 from neutron-org/fix_tick_fee_validation

Ensure deposit, withdraw and place/withdraw limit order correctly val…

pr0n00gler committed:
Merge branch 'update-sdk47' into merge_duality_neutron

pr0n00gler committed:
bump pfm to the latest release

pr0n00gler committed:
update whitelisted msgs

pr0n00gler committed:
fmt

Commit:
Update to pfm v7.1.0

pr0n00gler committed:
Merge pull request #363 from neutron-org/chore/bump-pfm

Chore: bump pfm

pr0n00gler committed:
Merge branch 'merge_duality_neutron' into fix/ibc-_swap_pfm_issues

pr0n00gler committed:
Merge pull request #359 from neutron-org/fix/ibc-_swap_pfm_issues

Audit FIX: ibc swap pfm issues

Commit:
Merge remote-tracking branch 'origin/merge_duality_neutron' into fix_deposit_spread

Commit:
Fix test logic to account for PFM changes

pr0n00gler commented:
https://github.com/neutron-org/neutron-tests/actions/runs/6865972417

Commit:
Return unused tokenIn to original creator addr

pr0n00gler committed:
Merge branch 'update-sdk47' into merge_duality_neutron

pr0n00gler committed:
Merge branch 'update-sdk47' into merge_duality_neutron

pr0n00gler committed:
Merge branch 'update-sdk47' into merge_duality_neutron

pr0n00gler committed:
whitelist dex stargate queries

pr0n00gler committed:
/neutron.dex.Query/LimitOrderTrancheAll query

Commit:
Revert "Merge pull request #359 from neutron-org/fix/ibc-_swap_pfm_issues"

This reverts commit 798bb43a453038f8a9b0c1dfb8bcf3731fd851e3, reversing
changes made to 483f875f12d28b3aaaf304194cc926de0fccf0c4.

Commit:
Revert "Merge pull request #369 from neutron-org/chore/bump-pfm-again"

This reverts commit 3765f6e71870c8f0f1b3f163ef5e8a634351c4c9, reversing
changes made to 3ee6217d485f6e818b2db7c9ad94869d1b8e0eba.

Commit:
Revert "bump pfm to the latest release"

This reverts commit 1ffecd060b4306212f5c6c011d80a80c0864ec31.

Commit:
Add validation for ibcSwap to ensure transfer receiver is sender

also ensure that swap amount <= transferAmount

Commit:
Fix tests

Commit:
fix ibcswap tests to account for changes to swap rounding

pr0n00gler committed:
whitelist ccvconsumer params

pr0n00gler committed:
Merge pull request #371 from neutron-org/fix/whitelist-dex-stargate-queries

Fix: whitelist dex stargate queries

pr0n00gler committed:
pfm update params whitelits

pr0n00gler committed:
Merge branch 'update-sdk47' into merge_duality_neutron

pr0n00gler committed:
fix tokenfactory params upgrade

pr0n00gler committed:
Merge pull request #344 from neutron-org/fix_deposit_spread

Audit Fix: prevent depositor or LP from creating unfair spreads

Commit:
Merge branch 'merge_duality_neutron' into fix/revert-pfm-upgrade

Commit:
Revert "pfm update params whitelits"

This reverts commit 196b157820d9a7490bb5b72fba231776b936a62b.

Commit:
whitelist pfm params

pr0n00gler committed:
Merge pull request #372 from neutron-org/fix/revert-pfm-upgrade

Fix/revert pfm upgrade

pr0n00gler committed:
Merge branch 'merge_duality_neutron' into chore/duality-proto-camelCase->snake_case

pr0n00gler committed:
Merge pull request #358 from neutron-org/chore/duality-proto-camelCase->snake_case

camelCase -> snake_case

Commit:
Revert "Merge pull request #372 from neutron-org/fix/revert-pfm-upgrade"

This reverts commit 4e974e456c7f465b8e86d2ba631a9d7549cbd5e6, reversing
changes made to 22e763ca031d2f0504d8459bc04acc95d48bfc2b.

Commit:
disable use of ibcswap middleware with forward-middleware

Commit:
Disable gmp swap and forward tests.

Add simple gmp ibcswap test

Commit:
Add test to ensure that swap-and-forward fails gracefully

Commit:
fix typo

Commit:
remove unused NonRefundable key from SwapMetadata

pr0n00gler committed:
Merge branch 'update-sdk47' into merge_duality_neutron

pr0n00gler committed:
Merge branch 'merge_duality_neutron' into fix/disable-pfm-revert-reverts

pr0n00gler committed:
lint && format

Commit:
delete commented code

Commit:
bump pfm to 7.1.2

pr0n00gler committed:
Merge branch 'update-sdk47' into merge_duality_neutron

pr0n00gler committed:
Merge branch 'merge_duality_neutron' into fix/disable-pfm-revert-reverts

pr0n00gler committed:
fix go.sum

pr0n00gler committed:
update linter in actions

pr0n00gler committed:
bump pfm

pr0n00gler committed:
update contracts

pr0n00gler committed:
Merge branch 'update-sdk47' into merge_duality_neutron

pr0n00gler committed:
Merge branch 'merge_duality_neutron' into fix/disable-pfm-revert-reverts

pr0n00gler commented:
https://github.com/neutron-org/neutron-tests/actions/runs/6990462605

pr0n00gler committed:
Merge pull request #374 from neutron-org/fix/disable-pfm-revert-reverts

revert PFM reversions. Disable PFM for ibcswap
